sdk/core/Azure.Core/src/Serialization/ModelJsonConverter.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Azure.Core.Serialization
+{
+    /// <summary>
+    /// A generic converter which allows <see cref="JsonSerializer"/> to be able to serialize and deserialize any models that implement <see cref="IModelJsonSerializable{T}"/>.
+    /// </summary>
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+#if !NET5_0 // RequiresUnreferencedCode in net5.0 doesn't have AttributeTargets.Class as a target, but it was added in net6.0
+    [RequiresUnreferencedCode("The constructors of the type being deserialized are dynamically accessed and may be trimmed.")]
+#endif
+    public class ModelJsonConverter : JsonConverter<IModelJsonSerializable<object>>
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+    {
+        /// <summary>
+        /// Gets the <see cref="ModelSerializerOptions"/> used to serialize and deserialize models.
+        /// </summary>
+        public ModelSerializerOptions ModelSerializerOptions { get; }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ModelJsonConverter"/> with a default format of <see cref="ModelSerializerFormat.Json"/>.
+        /// </summary>
+        public ModelJsonConverter()
+            : this(ModelSerializerFormat.Json) { }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ModelJsonConverter"/>.
+        /// </summary>
+        /// <param name="format"> The format to serialize to and deserialize from. </param>
+        public ModelJsonConverter(ModelSerializerFormat format)
+            : this(ModelSerializerOptions.GetOptions(format)) { }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ModelJsonConverter"/>.
+        /// </summary>
+        /// <param name="options">The <see cref="ModelSerializerOptions"/> to use.</param>
+        public ModelJsonConverter(ModelSerializerOptions options)
+        {
+            ModelSerializerOptions = options;
+        }
+
+        /// <inheritdoc/>
+        public override bool CanConvert(Type typeToConvert)
+        {
+            return !Attribute.IsDefined(typeToConvert, typeof(JsonConverterAttribute));
+        }
+
+        /// <inheritdoc/>
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+        public override IModelJsonSerializable<object> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+        {
+            using JsonDocument document = JsonDocument.ParseValue(ref reader);
+            return (IModelJsonSerializable<object>)ModelSerializer.Deserialize(BinaryData.FromString(document.RootElement.GetRawText()), typeToConvert, ModelSerializerOptions);
```

Comment:
Is there a way to do this all in UTF-8 bytes without transcoding through strings?

Reply:
Not that I have found unfortunately.

Reply:
I believe the following will prevent transcoding, but I don't know how that trades off perf against allocating a stream and writer:
```csharp
    using JsonDocument document = JsonDocument.ParseValue(ref reader);
    using MemoryStream stream = new();
    using Utf8JsonWriter writer = new(stream);
    document.WriteTo(writer);
    writer.Flush();
    stream.Position = 0;
    BinaryData data = BinaryData.FromStream(stream);
    return (IModelJsonSerializable<object>)ModelSerializer.Deserialize(data, typeToConvert, ModelSerializerOptions);
```

Reply:
yeah this a decent perf hit.  I think we will stick with what we have and we have the [follow up issue to finish the utf8jsonreader](https://github.com/Azure/azure-sdk-for-net/issues/38079) direct parsing which will give us a big perf boost.

```
Before
|                      Type |                                   Method |         Mean |      Error |     StdDev |   Gen0 | Allocated |
|-------------------------- |----------------------------------------- |-------------:|-----------:|-----------:|-------:|----------:|
|  AvailabilitySetDataModel | Deserialize_Utf8JsonReaderFromBinaryData |     3.039 us |  0.0143 us |  0.0127 us | 0.0243 |   1.42 KB |
| ResourceProviderDataModel | Deserialize_Utf8JsonReaderFromBinaryData | 2,061.217 us | 40.9207 us | 45.4833 us |      - | 770.94 KB |

After
|                      Type |                                   Method |         Mean |      Error |     StdDev |   Gen0 | Allocated |
|-------------------------- |----------------------------------------- |-------------:|-----------:|-----------:|-------:|----------:|
|  AvailabilitySetDataModel | Deserialize_Utf8JsonReaderFromBinaryData |     4.999 us |  0.0324 us |  0.0303 us | 0.0401 |   2.44 KB |
| ResourceProviderDataModel | Deserialize_Utf8JsonReaderFromBinaryData | 2,893.609 us | 47.8416 us | 44.7511 us |      - | 947.97 KB |
```